### PR TITLE
OPEN-10: Refactor SparqlService

### PR DESCRIPTION
**Pull request by MerlijnCuypers**
Refactor of SparqlService:
* Service is now a singleton
* Curl syntax is replaced by Guzzle
* Endpoints are now bound to the Guzzle Client obj, that can be done and dynamically changed in the setClient method
* when no special connection settings are send, the defaults will be used from the .env file
* Authentication is used for all queries (even select)
* Connection test checks or endpoint contains correct Sparql protocol
* queries can be triggered with get and post methods (deprecated performSparqlQuery still works)
* lastResponseCode property is available
* PHPunitTests are written for validation testing and a CRUD scenario

### CODE REVIEW

File: app/Services/SparqlService.php [hunk 1]
```diff
@@ -31,41 +37,111 @@ class SparqlService
...
+        $this->password = $password ?: env('SPARQL_WRITE_ENDPOINT_PASSWORD');
+        $this->defaultGraph = $defaultGraph ?: env('SPARQL_WRITE_GRAPH');
+
+        $this->setClient($endpoint ?: env('SPARQL_WRITE_ENDPOINT'));
+    }
+
+    /**
+     * Create new Guzzle client with fixed base_uri endpoint
+     *
+     * Is public as perhaps onther endpoints can be needed in logic
```
Comment (Jelle-S) [inline app/Services/SparqlService.php]: s/onther/other/

File: app/Services/SparqlService.php [hunk 2]
```diff
@@ -31,41 +37,111 @@ class SparqlService
...
+     * As currently digest authenitcation is currently only supported when using the cURL handler
```
Comment (Jelle-S) [inline app/Services/SparqlService.php]: currently [...] currently

File: app/Services/SparqlService.php [hunk 3]
```diff
@@ -31,41 +37,111 @@ class SparqlService
...
+     * http://docs.guzzlephp.org/en/stable/request-options.html#auth
+     *
+     * The base baseConnectionTest is triggered here to check or Client is correctly set
+     * When not correct, the thrown error will prevent any further logic being send to the objedt
```
Comment (Jelle-S) [inline app/Services/SparqlService.php]: s/objedt/object/

File: app/Services/SparqlService.php [hunk 4]
```diff
@@ -31,41 +37,111 @@ class SparqlService
...
+        $options['handler'] = HandlerStack::create($handler); // Wrap w/ middleware
+        $options['base_uri'] = $endpoint;
+        $this->guzzleClient = new Client($options);
+
+        $this->baseConnectionTest();
+
+    }
+
+    /**
+     * Test or the current Guzzle client contains the Sparql protocol
```
Comment (Jelle-S) [inline app/Services/SparqlService.php]: s/or/if/

File: app/Services/SparqlService.php [hunk 5]
```diff
@@ -74,148 +150,123 @@ public function performSparqlQuery($query, $method = 'GET', $format = 'turtle')
...
-     * Prepare the query statement to be passed in a cURL requests
-     * meaning we need to apply proper url encoding and take into
-     * account specific characters
+     * Prepare and perform a POST to the SparQL
      *
-     * @param  string $query
-     * @return string
+     * Reset of $this->lastResponseCode
+     * Assemble the query for POST syntax
+     * As graph-uri value, the efaultGraph is set
```
Comment (Jelle-S) [inline app/Services/SparqlService.php]: s/efaultGraph/defaultGraph/

File: app/Services/SparqlService.php [hunk 6]
```diff
@@ -74,148 +150,123 @@ public function performSparqlQuery($query, $method = 'GET', $format = 'turtle')
...
-            $message = "The SPARQL endpoint returned a 500 error. If the SPARQL query contained a parameter, don't forget to pass them as a query string parameter. The URI was: $uri";
-
-            throw new \Exception($message);
-        }
+        return $this->executeQuery('GET', $uri);
+    }
 
-        curl_close($curl);
+    /**
+     * transform qQuery string to SparQL fit for uri
```
Comment (Jelle-S) [inline app/Services/SparqlService.php]: s/transform/Transform/
s/qQuery/query/

File: app/Services/SparqlService.php [hunk 7]
```diff
@@ -74,148 +150,123 @@ public function performSparqlQuery($query, $method = 'GET', $format = 'turtle')
...
-        return $response;
+        return $query;
     }
 
     /**
-     * Make and return the request URI based on
-     * the passed query and the configured endpoint
+     * Perform the query to the SPARQL endpoint
+     *
+     * Standard use auht. This is not needed for the GET SELECT,
```
Comment (Jelle-S) [inline app/Services/SparqlService.php]: s/auht/auth/

File: app/Services/SparqlService.php [hunk 8]
```diff
@@ -74,148 +150,123 @@ public function performSparqlQuery($query, $method = 'GET', $format = 'turtle')
...
+     * but to make extra logic for a "less safe exception" is ... not logic
      *
+     * 'application/sparql-query' is set es Content-type
```
Comment (Jelle-S) [inline app/Services/SparqlService.php]: s/es/as/

File: app/Services/SparqlService.php [hunk 9]
```diff
@@ -74,148 +150,123 @@ public function performSparqlQuery($query, $method = 'GET', $format = 'turtle')
...
+     * @param  array  $options
+     * @return Psr\Http\Message\StreamInterface
      */
-    private function makeGetRequestUri($query, $format = null)
+    private function executeQuery($verb, $query, $options = [])
     {
-        $requestUri = $this->endpoint . '?query=' . $query;
+        $options['auth'] = [$this->username, $this->password, 'digest'];
+        $options['Content-Type'] = 'application/sparql-query';
+        // try {
```
Comment (Jelle-S) [inline app/Services/SparqlService.php]: remove comment (or implement try catch if needed)

File: app/Services/SparqlService.php [hunk 10]
```diff
@@ -74,148 +150,123 @@ public function performSparqlQuery($query, $method = 'GET', $format = 'turtle')
...
+        $this->lastResponseCode = $response->getStatusCode();
+        // Virtuoso documentation states it only returns 200, 400, 500 status codes
+        // but apparently they mean 2xx, 4xx and 5xx
+        if ($response->getReasonPhrase() != 'OK' && $this->lastResponseCode > 299) {
+            $message = "Something went wrong while executing query.";
+            throw new \Exception($message);
+        }
+
+        if ($this->lastResponseCode >= '400' && $this->lastResponseCode <= '500') {
+            $message = "The SPARQL endpoint returned an error.";
```
Comment (Jelle-S) [inline app/Services/SparqlService.php]: Does the $response contain the error message? If so, append it to the exception message, so we know what's going on.

File: app/Services/SparqlService.php [hunk 11]
```diff
@@ -74,148 +150,123 @@ public function performSparqlQuery($query, $method = 'GET', $format = 'turtle')
...
+            throw new \Exception($message);
+        }
 
-        if (! empty($format)) {
-            $requestUri .= '&format=' . $format;
+        if ($this->lastResponseCode == '500') {
+            $message = "The SPARQL endpoint returned a 500 error.";
```
Comment (Jelle-S) [inline app/Services/SparqlService.php]: Same remark about the error message.

File: app/Services/SparqlService.php (hunk 10 above)
Comment (MerlijnCuypers) [reply app/Services/SparqlService.php]: No: you get the error message from Guzzle what has nothing to do with sparql error.
The return is html from the website but not something we can log

File: app/Services/SparqlService.php (hunk 11 above)
Comment (MerlijnCuypers) [reply app/Services/SparqlService.php]: same reply here

File: app/Services/SparqlService.php (hunk 10 above)
Comment (Jelle-S) [reply app/Services/SparqlService.php]: Too bad, but nothing we can do about it then.

File: app/Services/SparqlService.php [hunk 12]
```diff
@@ -74,148 +150,122 @@ public function performSparqlQuery($query, $method = 'GET', $format = 'turtle')
...
+     * @return Psr\Http\Message\StreamInterface
      */
-    private function makePostRequestUri($query, $format = null)
+    private function handleResponse(Response $response)
     {
-        $requestUri = $this->endpoint . '?query=' . $query;
+        $this->lastResponseCode = $response->getStatusCode();
+        // Virtuoso documentation states it only returns 200, 400, 500 status codes
+        // but apparently they mean 2xx, 4xx and 5xx
+        if ($response->getReasonPhrase() != 'OK' && $this->lastResponseCode > 299) {
```
Comment (zero2one) [inline app/Services/SparqlService.php]: !==

**Pull request merged**